instaclustr/resource_kafka_topic.go

```diff
@@ -0,0 +1,382 @@
+		return fmt.Errorf("[Error] Error in getting the status of the cluster: %s", err)
+	}
+	if cluster.ClusterStatus != "RUNNING" {
+		return fmt.Errorf("[Error] Cluster %s is not RUNNING. Currently in %s state", cluster_id, cluster.ClusterStatus)
+	}
+	createKafkaTopicRequest := CreateKafkaTopicRequest{
+		Topic:             topic,
+		ReplicationFactor: d.Get("replication_factor").(int),
+		Partitions:        d.Get("partitions").(int),
+	}
+	var jsonStr []byte
+	jsonStr, err = json.Marshal(createKafkaTopicRequest)
+	if err != nil {
+		return fmt.Errorf("[Error] Error creating kafka topic creation request: %s", err)
+	}
+
+	err = client.CreateKafkaTopic(cluster_id, jsonStr)
+	if err != nil {
+		return fmt.Errorf("[Error] Error creating kafka topic: %s", err)
+	}
+
+	d.SetId(fmt.Sprintf("%s&%s", cluster_id, topic))
+
+	log.Printf("[INFO] Kafka topic %s has been created.", topic)
+	return nil
+}
+
+func resourceKafkaTopicRead(d *schema.ResourceData, meta interface{}) error {
+	cluster_id := d.Get("cluster_id").(string)
+	topic := d.Get("topic").(string)
+	client := meta.(*Config).Client
+
+	// Cluster has to reach running state first
+	cluster, err := client.ReadCluster(cluster_id)
+	if err != nil {
+		return fmt.Errorf("[Error] Error in getting the status of the cluster: %s", err)
+	}
+	if cluster.ClusterStatus != "RUNNING" {
+		return fmt.Errorf("[Error] Cluster %s is not RUNNING. Currently in %s state", cluster_id, cluster.ClusterStatus)
+	}
+	log.Printf("[INFO] Reading the config of topic %s.", topic)
+	kafkaTopicConfig, err := client.ReadKafkaTopicConfig(cluster_id, topic)
+	if err != nil {
+		return fmt.Errorf("[Error] Error reading Kafka topic %s's config: %s", topic, err)
+	}
+
+	// It won't read the topic's config if the config{} block is missing in the resource
+	if len(d.Get("config").([]interface{})) == 0 {
+		log.Printf("[INFO] A block of config{} needs to be provided in the resource to read the topic's config.")
+		log.Printf("[INFO] Topic %s's config is not read.", topic)
+		return nil
+	}
+
+	config := make(map[string]interface{}, 0)
+	configList := make([]map[string]interface{}, 0)
+	err = mapstructure.Decode(kafkaTopicConfig.Config, &config)
+	if err != nil {
+		return err
+	}
+	configList = append(configList, config)
+	err = d.Set("config", configList)
+	if err != nil {
+		return err
+	}
+
+	log.Printf("[INFO] Successfully fetch the config of topic %s.", topic)
+	return nil
+}
+
+// There are two use cases of resourceKafkaTopicUpdate:
+// 1. Providing all the possible configs in resource, and modify the ones you want to update. Then keeping all the
+// configs in the resource.
+// 2. Providing some configs (not all) you want to update, and then keep only these configs in the resource.
+// Terraform always detects that there is an update in case 2 if you run the plan again, because resourceKafkaTopicRead
+// reads all the configs, but there are only some of them provided in resource. For the missing configs, Terraform just
```

Comment:
If the default values line up with the API's defaults, shouldn't terraform detect no change?

Reply:
There are no API's defaults, it will read the current configs for a topic. And it depends whether there is a full list of topic config provided in the resource. Like I said in the comment, if users only provide a partial list of config in the resource, Terraform will detect other missing configs, and set them as null.